tests/python/pants_test/backend/jvm/tasks/jvm_compile/java/test_zinc_compile_integration.py

```diff
@@ -254,3 +279,105 @@ def test_hermetic_binary_with_dependencies(self):
         ]:
           path = os.path.join(compile_dir, path_suffix)
           self.assertTrue(os.path.exists(path), "Want path {} to exist".format(path))
+
+  def test_hermetic_binary_cache_with_dependencies(self):
+    file_abs_path = os.path.join(get_buildroot(),
+      'examples/src/scala/org/pantsbuild/example/hello/exe/Exe.scala')
+
+    def run_then_overwrite_file_then_run_again(file_path):
```

Comment:
I think this may be a little more clear as:

```
do initial run...
with temporarily_overwritten_file_content(file_abs_path, new_temp_target):
  do next run...
```

which I think should work if instead of taking a `method_to_run` and calling it, you do the file write, and then just `yield` where you currently have `method_to_run(file_path)`